contracts/libraries/VaultLib.sol

```diff
@@ -266,9 +269,50 @@ library VaultLibrary {
 
         self.vault.lv.issue(from, amount);
 
+        self.vault.userLvBalance[from].balance += amount;
         received = amount;
     }
 
+    // returns the amount of PA user able to redeem
+    function redeemablePA(State storage self, address user) external view returns (uint256) {
+        return self.vault.userLvBalance[user].balance;
+    }
+
+    // returns the amount of PA as per price of PA with LV and given LV amount
+    // lv price = paAvailable/lvTotalSupply
+    function _calculatePaPriceForLv(State storage self, uint256 lvAmt) internal view returns (uint256 paAmount) {
+        return lvAmt * ERC20(self.info.pair0).balanceOf(address(this)) / ERC20(self.vault.lv._address).totalSupply();
+    }
+
+    // preview a redeem action for PA with up to current amount of LV,
+    // returns the amount of PA that user will receive based on his LV deposited
+    function previewRedeemPaWithLv(State storage self, address user, uint256 amount) external view returns (uint256) {
+        uint256 userLvAmt = self.vault.userLvBalance[user].balance;
+        if (userLvAmt > amount) {
+            return amount;
+        } else {
+            return _calculatePaPriceForLv(self, amount);
+        }
+    }
+
+    // redeem LV for PA with up to current amount of LV,
+    // returns the PA user redeemed with his LV amount
+    function redeemPaWithLv(State storage self, address user, uint256 amount) external returns (uint256) {
```

Comment:
we should remove the amount and withdraw all of user balance

Reply:
this is also missing a flag that must be set after every expiry to prevent user withdrawing more than once per expiry